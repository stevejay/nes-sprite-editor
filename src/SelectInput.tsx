import React from "react";
import styles from "./SelectInput.module.scss";

type Props = {
  options: Array<{ id: string; label: string }>;
  selectedId: string;
  onChange: (id: string) => void;
  className?: string;
};

const SelectInput: React.FunctionComponent<Props> = ({
  options,
  selectedId,
  onChange,
  className = ""
}) => (
  <div className={`${styles.container} ${className}`}>
    <select onChange={event => onChange(event.target.value)}>
      {options.map(option => (
        <option value={option.id} selected={selectedId === option.id}>
          {option.label}
        </option>
      ))}
    </select>
  </div>
);

export default SelectInput;
