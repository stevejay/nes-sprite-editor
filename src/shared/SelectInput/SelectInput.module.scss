@import "styles/_base.scss";

$button-v-padding: 10;

.container {
  position: relative;
  color: inherit;
  width: 100%;

  &::after {
    content: "\25BE";
    position: absolute;
    top: 50%;
    transform: translateY(-50%);
    right: rem(13);
    color: $md-grey-800;
    pointer-events: none;
  }

  &.disabled::after {
    color: $md-grey-500;
  }
}

.container > select {
  -webkit-appearance: none;
  -moz-appearance: none;
  width: 100%;
  height: rem($button-v-padding + $button-v-padding + 16 + 2);
  padding: rem($button-v-padding) rem(32) rem($button-v-padding) rem(12);
  font-family: inherit;
  font-size: rem(16);
  line-height: 1;
  vertical-align: middle;
  border: 1px solid $md-grey-500;
  border-radius: rem(3);
  outline: 0;
  background-color: $md-white;
  line-height: 1.5rem;
  padding-top: 0;
  padding-bottom: 0;

  &:disabled {
    background-color: $md-grey-200;
    border-color: $md-grey-400;
    cursor: not-allowed;
  }
}

.container > select:-moz-focusring,
.container > select::-moz-focus-inner {
  color: transparent !important;
  text-shadow: 0 0 0 #000 !important;
  background-image: none !important;
  border: 0;
}

.container > select:focus {
  @include focus-ring;
}
