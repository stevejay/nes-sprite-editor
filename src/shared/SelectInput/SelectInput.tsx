import React from "react";
import styles from "./SelectInput.module.scss";
import { isEmpty, isNil } from "lodash";
import classNames from "classnames";

type Option<IdT> = {
  id: IdT;
  label: string;
};

type Props<IdT> = {
  id?: string;
  options: Array<Option<IdT>>;
  value: IdT | null;
  disabled?: boolean;
  onChange: (id: IdT) => void;
  className?: string;
};

const SelectInput = <P extends string | number>({
  id,
  options,
  value,
  disabled = false,
  onChange,
  className = ""
}: Props<P>) => {
  const showDisabled = disabled || isNil(value) || isEmpty(options);

  const containerClassNames = classNames(styles.container, className, {
    [styles.disabled]: showDisabled
  });

  return (
    <div className={containerClassNames}>
      <select
        id={id}
        value={value || undefined}
        onChange={event => onChange(event.target.value as P)}
        disabled={showDisabled}
        tabIndex={0}
      >
        {options.map(option => (
          <option key={option.id} value={option.id}>
            {option.label}
          </option>
        ))}
      </select>
    </div>
  );
};

export default SelectInput;
