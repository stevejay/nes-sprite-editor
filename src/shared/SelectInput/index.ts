export { default } from "./SelectInput";
import { Option as SelectInputOption } from "./SelectInput";
export type Option<IdT> = SelectInputOption<IdT>;
