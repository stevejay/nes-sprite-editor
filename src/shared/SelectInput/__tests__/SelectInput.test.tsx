import React from "react";
import { render, cleanup } from "react-testing-library";
import SelectInput from "../SelectInput";
import { noop } from "lodash";

afterEach(cleanup);

const options = [
  { id: 1, label: "Some text 1" },
  { id: 2, label: "Some text 2" },
  { id: 3, label: "Some text 3" }
];

test("displays correctly", async () => {
  const { container } = render(
    <SelectInput<number>
      id="the-id"
      name="the-name"
      options={options}
      value={2}
      disabled={false}
      onChange={noop}
    />
  );
  expect(container).toMatchSnapshot();
});
