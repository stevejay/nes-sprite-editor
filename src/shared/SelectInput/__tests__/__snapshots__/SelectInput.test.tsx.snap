// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays correctly 1`] = `
<div>
  <div
    class="container"
  >
    <select
      id="the-id"
      name="the-name"
      tabindex="0"
    >
      <option
        value="1"
      >
        Some text 1
      </option>
      <option
        value="2"
      >
        Some text 2
      </option>
      <option
        value="3"
      >
        Some text 3
      </option>
    </select>
  </div>
</div>
`;
